lightningd/invoice.c

```diff
@@ -1248,7 +1250,7 @@ static void json_add_invoices(struct json_stream *response,
 		}
 	} else {
 		memset(&it, 0, sizeof(it));
-		while (wallet_invoice_iterate(wallet, &it)) {
+		while (wallet_invoice_iterate(wallet, &it, listindex, liststart)) {
```

Comment:
Mh! we always avoid putting specific fields for specific fields, and I think this was the right choice?

Why we do not steal your idea of `filtering` but inject the field inside the JSON RPC params array?

I wrote an I think modular version of the paginator here https://github.com/vincenzopalazzo/lightning/commit/7401722f3304a2ad8885b5c2ae72519abb80cc13#diff-401ba4b6734345a7b89101b868e5709aa58d0e6d6e33b6df2451bbfb488e087f by stealing your idea of `filter` filed.

Reply:
There are several things here:

1. We need to drive the offset and limit all the way into the db code for efficiency.
2. We need to have a well-defined stable order (this patch defines two in fact, creation and update order) otherwise pagination can occasionally miss entries.

Now, we could hide these, as we did for filtering, but I don't think that's a good idea, since they only apply to a small subset of commands, and you probably really care if they are ignored!

We are unlikely to implement generic output selection, simply because doing it efficiently means we have to have a powerful db engine which maintains a huge number of indexes for each possible query. Hence I expect all usage in practice will be either a single "key" query, or built by keeping your own index and using `wait` to detect changes, additions or deletions.

Reply:
>We are unlikely to implement generic output selection, simply because doing it efficiently means we have to have a powerful db engine which maintains a huge number of indexes for each possible query. Hence I expect all usage in practice will be either a single "key" query, or built by keeping your own index and using wait to detect changes, additions or deletions.

Yeah, I agree I was worried into abuse much of this dynamic filtering, but I think we need some way to decide a runtime if I want the result in reverse order or filtering by timestamp. 

I had a use case where I would like to have the forwarding by channel id or by timestamp, and maybe have a way to query commands like `listnode` and `listpeer` in a `batch`  way! like `listnodes batch=[".....", "...."]` 

>We need to have a well-defined stable order (this patch defines two in fact, creation and update order) otherwise pagination can occasionally miss entries.  

I guess that I am missing the case where we miss an entry but maybe I am missing just the use case where this happens.